on_device_training/training_api_demo/mnist_training_example.ipynb

```diff
@@ -0,0 +1,473 @@
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Serving 'data/eval_model.onnx' at http://localhost:8081\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "('localhost', 8081)"
+      ]
+     },
+     "execution_count": 65,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "netron.start(\"data/eval_model.onnx\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![](https://i.imgur.com/C7W7cw2.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### After generating the required files for training, let's start training out model.\n",
+    "##### Let's load the datasets from torchvision."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 66,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "batch_size = 64\n",
+    "train_kwargs = {'batch_size': batch_size}\n",
+    "test_batch_size = 1000\n",
+    "test_kwargs = {'batch_size': test_batch_size}\n",
+    "\n",
+    "transform = transforms.Compose([\n",
+    "    transforms.ToTensor(),\n",
+    "    transforms.Normalize((0.1307,), (0.3081,))\n",
+    "    ])\n",
+    "\n",
+    "dataset1 = datasets.MNIST('../data', train=True, download=True,\n",
+    "                    transform=transform)\n",
+    "dataset2 = datasets.MNIST('../data', train=False,\n",
+    "                    transform=transform)\n",
+    "train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)\n",
+    "test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Use the saved files to initialize the state, model and optimizer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 67,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create checkpoint state.\n",
+    "state = CheckpointState(\"data/checkpoint.ckpt\")\n",
+    "\n",
+    "# Create module.\n",
+    "model = Module(\"data/training_model.onnx\", state, \"data/eval_model.onnx\")\n",
+    "\n",
+    "# Create optimizer.\n",
+    "optimizer = Optimizer(\"data/optimizer.onnx\", model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Training and Testing Loops definition."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 76,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Util function to convert logits to predictions.\n",
+    "def get_pred(logits):\n",
+    "    return np.argmax(logits, axis=1)\n",
+    "\n",
+    "# Training Loop :\n",
+    "def train(epoch):\n",
+    "    model.train()\n",
+    "    losses = []\n",
+    "    for batch_idx, (data, target) in enumerate(train_loader):\n",
+    "        forward_inputs = [data.reshape(len(data),784).numpy(),target.numpy().astype(np.int32)]\n",
```

Comment:
why do we need this reshape?

Reply:
because the data comes as 28x28 from datasets, and the input of the model is 784